src/Oscoin/Storage/Block/Abstract.hs

```diff
@@ -7,37 +7,59 @@
 -}
 
 module Oscoin.Storage.Block.Abstract
-    ( BlockStore(..)
+    ( BlockStoreReader(..)
+    , BlockStore
+    , GetBlocksQuery(..)
 
-    , hoistBlockStore
-    , noValidation
+    , hoistBlockStoreReader
     , insertBlocksNaive
     , isNovelBlock
+
+    -- * Internals
+    , BlockStoreWriter(..)
+    , hoistBlockStoreWriter
     ) where
 
 import           Oscoin.Prelude
 
-import           Oscoin.Consensus.Types (Validate)
 import           Oscoin.Crypto.Blockchain (TxLookup)
 import           Oscoin.Crypto.Blockchain.Block
 import           Oscoin.Crypto.Hash (Hashed)
+import           Oscoin.Time.Chrono (NewestFirst, OldestFirst)
+
+-- The full 'BlockStore' API is the composition of its public and private parts.
+type BlockStore c tx s m =
+    ( BlockStoreReader c tx s m
+    , BlockStoreWriter c tx s m
+    )
 
--- | An handle over a block storage backend.
-data BlockStore c tx s m = BlockStore
-    { scoreBlock      :: Block c tx (Sealed c s) -> Score
-    -- ^ When given a 'Block', return its 'Score'.
-    , insertBlock     :: Block c tx (Sealed c s) -> m ()
+data GetBlocksQuery c =
+      ByDepth Depth
+    -- ^ Get a collection of blocks from the input 'Depth' up to the tip.
+    | ByRootHash (BlockHash c)
```

Comment:
I think the usage of `Root` here may be confusing (since the root of the block tree would actually be the genesis) - maybe `ByParentHash` or `ByBlockHash`, depending on which one you mean?

Reply:
More general comment - is there a reason why we don't just have two functions: `getBlocksByDepth` and `getBlocksByParentHash`? It's a lot simpler if there isn't a strong reason for the extra type.

Reply:
No strong reason if not API economy -- I thought this could belong to a set of related query on the chain, but perhaps you are right, this is premature optimisation and actually clutters a bit the API.

Reply:
Yeah I think it introduces "yet another type" I need to understand to use the function.